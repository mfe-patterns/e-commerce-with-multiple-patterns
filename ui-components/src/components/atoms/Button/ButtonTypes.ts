export enum ButtonAction {
  PRIMARY = 'primary',
  SECONDARY = 'secondary',
}

export enum ButtonSize {
  DEFAULT = 'large',
  XLARGE = 'xlarge',
  MEDIUM = 'medium',
  SMALL = 'small',
  XSMALL = 'xsmall',
}

export enum ButtonSpinnerSize {
  large = 20,
  xlarge = 24,
  medium = 19,
  small = 18,
  xsmall = 16,
}
