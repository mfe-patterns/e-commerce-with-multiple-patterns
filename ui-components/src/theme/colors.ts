interface Shade {
  50: string;
  100: string;
  200: string;
  300: string;
  400: string;
  500: string;
  600: string;
  700: string;
}

const dark: Shade = {
  50: "#11111193",
  100: "#111111a2",
  200: "#111111B3",
  300: "#111111B9",
  400: "#111111CC",
  500: "#111111d8",
  600: "#111111eb",
  700: "#111111",
};
const neutral: Shade = {
  50: "#fafafa",
  100: "#f2f2f2",
  200: "#ededed",
  300: "#dedede",
  400: "#cccccc",
  500: "#666666",
  600: "#434343",
  700: "#343434",
};

const primary: Shade = {
  50: "#e3f2fd",
  100: "#bbdefb",
  200: "#90caf9",
  300: "#64b5f6",
  400: "#42a5f5",
  500: "#2196f3",
  600: "#1e88e5",
  700: "#1976d2",
};

const white = "#ffffff";

export interface IColorPallete {
  neutral: Shade;
  dark: Shade;
  primary: Shade;
  white: string;
  black: string;
  error: string;
  success: string;
  text: string;
  transparent: string;
  link: string;
}

export const colors: IColorPallete = {
  neutral,
  dark,
  primary,
  white,
  black: "#000000",
  error: "#dc3545",
  success: "#28a745",
  text: dark[700],
  transparent: "transparent",
  link: "#0064cc",
};
