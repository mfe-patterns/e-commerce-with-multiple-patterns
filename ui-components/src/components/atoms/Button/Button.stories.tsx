import React from 'react';
import { storiesOf } from '@storybook/react';

import { Button } from './Button';
import { Wrapper } from './Button.styles';
import { ButtonAction, ButtonSize } from './ButtonTypes';

storiesOf('Atoms/Button', module)
  .add('Primary Button', () => (
    <div
      style={{
        flex: 1,
        width: '100%',
        maxWidth: '100%',
        flexWrap: 'wrap',
      }}>
      <Wrapper>
        <Button
          buttonType={ButtonAction.PRIMARY}
          onClick={() => { }}
          size={ButtonSize.XLARGE}
          disabled={false}>
          Primary Extra Large Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.PRIMARY}
          onClick={() => { }}
          size={ButtonSize.XLARGE}
          disabled>
          Primary Extra Large Disabled Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.PRIMARY}
          onClick={() => { }}
          size={ButtonSize.DEFAULT}
          disabled={false}>
          Primary Default (Large) Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.PRIMARY}
          onClick={() => { }}
          size={ButtonSize.MEDIUM}
          disabled={false}>
          Primary Medium Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.PRIMARY}
          onClick={() => { }}
          size={ButtonSize.SMALL}
          disabled={false}>
          Primary Small Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.PRIMARY}
          onClick={() => { }}
          size={ButtonSize.XSMALL}
          disabled={false}>
          Primary Extra Small Button
        </Button>
      </Wrapper>
    </div>
  ))
  .add('Secondary Button', () => (
    <div
      style={{
        flex: 1,
        width: '100%',
        maxWidth: '100%',
        flexWrap: 'wrap',
      }}>
      <Wrapper>
        <Button
          buttonType={ButtonAction.SECONDARY}
          onClick={() => { }}
          size={ButtonSize.XLARGE}
          disabled={false}>          
          Secondary Extra Large Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.SECONDARY}
          onClick={() => { }}
          size={ButtonSize.XLARGE}
          disabled>
          Secondary Extra Large Disabled Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.SECONDARY}
          onClick={() => { }}
          size={ButtonSize.DEFAULT}
          disabled={false}>
          Secondary Default (Large) Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.SECONDARY}
          onClick={() => { }}
          size={ButtonSize.MEDIUM}
          disabled={false}>
          Secondary Medium Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.SECONDARY}
          onClick={() => { }}
          size={ButtonSize.SMALL}
          disabled={false}>
          Secondary Small Button
        </Button>
      </Wrapper>
      <Wrapper>
        <Button
          buttonType={ButtonAction.SECONDARY}
          onClick={() => { }}
          size={ButtonSize.XSMALL}
          disabled={false}>
          Secondary Extra Small Button
        </Button>
      </Wrapper>            
    </div>
  ));