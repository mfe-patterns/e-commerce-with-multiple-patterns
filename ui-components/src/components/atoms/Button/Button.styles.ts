import styled, { css } from "styled-components";
import { ButtonAction, ButtonSize } from "./ButtonTypes";

interface IButton {
  children: React.ReactNode;
  buttonType: ButtonAction;
  size: ButtonSize;
  disabled: boolean;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Wrapper = styled.div`
  margin: 20px;
  width: 40%;
`;

export const Button = styled.button<IButton>`
  width: 100%;
  cursor: pointer;
  min-height: ${({ theme, size }) => theme.button.size[size].height};
  border: ${({ theme, buttonType }) => theme.button.type[buttonType].border};
  border-radius: ${({ theme, size }) => theme.button.size[size].borderRadius};
  justify-content: center;
  align-items: center;
  border-color: ${({ theme, buttonType }) =>
    theme.button.type[buttonType].borderColor};
  background-color: ${({ theme, buttonType }) =>
    theme.button.type[buttonType].backgroundColor};
  padding: ${({ theme, size }) =>
    `${theme.button.size[size].paddingVertical} ${theme.button.size[size].paddingHorizontal}`};
  ${({ buttonType, disabled, theme }) =>
    disabled &&
    css`
      border-color: ${theme.button.type[buttonType].disabled.borderColor};
      background-color: ${theme.button.type[buttonType].disabled
        .backgroundColor};
    `}
  color: ${({ theme, buttonType, disabled }) =>
    disabled
      ? theme.button.type[buttonType].disabled.color
      : theme.button.type[buttonType].color};
`;
