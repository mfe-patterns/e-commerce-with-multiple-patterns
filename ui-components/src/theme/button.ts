import { colors } from "./colors";

interface IButtonType {
  color: string;
  backgroundColor?: string;
  borderColor?: string;
  fontFamily: string;
  border: string;
  disabled: {
    color: string;
    backgroundColor?: string;
    borderColor?: string;
  };
  horizontalPadding: boolean;
}

export const primary: IButtonType = {
  color: colors.white,
  borderColor: colors.transparent,
  fontFamily: "Roboto-Regular",
  border: "none",
  disabled: {
    color: colors.white,
    borderColor: colors.transparent,
  },
  horizontalPadding: true,
};
export const secondary: IButtonType = {
  backgroundColor: colors.white,
  color: colors.text,
  fontFamily: "Roboto-Regular",
  border: "1px solid",
  disabled: {
    backgroundColor: colors.white,
    color: colors.dark[100],
  },
  horizontalPadding: true,
};

interface IButtonSize {
  height: string;
  borderRadius: string;
  fontSize: string;
  lineHeight: string;
  letterSpacing: string;
  paddingVertical: string;
  paddingHorizontal: string;
}

interface ISize {
  xlarge: IButtonSize;
  large: IButtonSize;
  medium: IButtonSize;
  small: IButtonSize;
  xsmall: IButtonSize;
}

const size: ISize = {
  xlarge: {
    height: "56px",
    borderRadius: "28px",
    fontSize: "16px",
    lineHeight: "24px",
    letterSpacing: "0.7px",
    paddingVertical: "15px",
    paddingHorizontal: "29.5px",
  },
  large: {
    height: "46px",
    borderRadius: "23px",
    fontSize: "15px",
    lineHeight: "20px",
    letterSpacing: "0.7px",
    paddingVertical: "12px",
    paddingHorizontal: "25.5px",
  },
  medium: {
    height: "40px",
    borderRadius: "20px",
    fontSize: "14px",
    lineHeight: "20px",
    letterSpacing: "0.7px",
    paddingVertical: "10px",
    paddingHorizontal: "21.5px",
  },
  small: {
    height: "36px",
    borderRadius: "18px",
    fontSize: "13px",
    lineHeight: "19px",
    letterSpacing: "0.7px",
    paddingVertical: "8px",
    paddingHorizontal: "20px",
  },
  xsmall: {
    height: "29px",
    borderRadius: "14.5px",
    fontSize: "12px",
    lineHeight: "19px",
    letterSpacing: "0.7px",
    paddingVertical: "5px",
    paddingHorizontal: "14px",
  },
};

export interface IButton {
  type: {
    primary: IButtonType;
    secondary: IButtonType;
  },
  size: ISize;
}

export const button: IButton = {
  type: {
    primary: {
      ...primary,
      disabled: {
        ...primary.disabled,
        backgroundColor: colors.primary[200],
      },
      backgroundColor: colors.primary[500],
    },
    secondary: {
      ...secondary,
      borderColor: colors.primary[500],
      disabled: {
        ...secondary.disabled,
        borderColor: colors.primary[200],
      },
    },
  },
  size,
};
