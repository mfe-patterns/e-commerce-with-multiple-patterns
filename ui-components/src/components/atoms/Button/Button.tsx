import React from 'react';

import * as S from './Button.styles';
import { ButtonAction, ButtonSize } from './ButtonTypes';

export interface IButton {
  children: React.ReactNode;
  buttonType?: ButtonAction;
  size?: ButtonSize;
  disabled?: boolean;
  onClick: (event: React.MouseEvent<HTMLButtonElement>) => void;
}

export const Button = (props: IButton) => {
  const {
    children,
    buttonType = ButtonAction.PRIMARY,
    size = ButtonSize.DEFAULT,
    disabled = false,
    onClick,
    ...rest
  } = props;

  return (
    <S.Button
      buttonType={buttonType}
      size={size}
      disabled={disabled}
      onClick={onClick}
      {...rest}>
      {children}
    </S.Button>
  );
};

export { ButtonAction, ButtonSize };
