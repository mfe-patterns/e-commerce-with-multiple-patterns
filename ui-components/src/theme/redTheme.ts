import { ITypography, typography } from "./typography";
import { IColorPallete, colors } from "./colors";
import { ITextInput, textInput } from "./textInput";
import { button, IButton, primary, secondary } from "./button";

const redPalette: IColorPallete = {
  ...colors,
  primary: {
    50: "#ffebee",
    100: "#ffcdd2",
    200: "#ef9a9a",
    300: "#e57373",
    400: "#ef5350",
    500: "#f44336",
    600: "#e53935",
    700: "#d32f2f",
  },
};

const redButton: IButton = {
  ...button,
  type: {
    primary: {
      ...primary,
      disabled: {
        ...primary.disabled,
        backgroundColor: redPalette.primary[200],
      },
      backgroundColor: redPalette.primary[500],
    },
    secondary: {
      ...secondary,
      borderColor: redPalette.primary[500],
      disabled: {
        ...secondary.disabled,
        borderColor: redPalette.primary[200],
      },
    },
  },
};

interface ITheme {
  typography: ITypography;
  colors: IColorPallete;
  textInput: ITextInput;
  button: IButton;
  gridSize: number;
}

export const theme: ITheme = {
  typography,
  colors: redPalette,
  textInput,
  button: redButton,
  gridSize: 8,
};
