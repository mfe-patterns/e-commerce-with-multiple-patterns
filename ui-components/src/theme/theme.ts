import { ITypography, typography } from "./typography";
import { IColorPallete, colors } from "./colors";
import { ITextInput, textInput } from "./textInput";
import { button, IButton } from "./button";

interface ITheme {
  typography: ITypography;
  colors: IColorPallete;
  textInput: ITextInput;
  button: IButton;
  gridSize: number;
}

export const theme: ITheme = {
  typography,
  colors,
  textInput,
  button,
  gridSize: 8
};
