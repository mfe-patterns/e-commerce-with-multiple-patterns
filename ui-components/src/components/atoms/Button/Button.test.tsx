import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import 'jest-styled-components/';

import { Button } from './Button';
import { ButtonAction, ButtonSize } from './ButtonTypes';
import { theme } from '../../../theme/theme';

describe('Button', () => {
  const onClick = jest.fn();
  test('should render primary default button', () => {
    const { container, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          buttonType={ButtonAction.PRIMARY}
          size={ButtonSize.DEFAULT}
          onClick={onClick}
          data-testid="primary-button">
          Click here
        </Button>
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
    fireEvent.click(getByTestId('primary-button'));
    expect(onClick).toHaveBeenCalled();
  });

  test('should render the primary button in disabled mode', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          buttonType={ButtonAction.PRIMARY}
          size={ButtonSize.DEFAULT}
          onClick={onClick}
          disabled
          data-testid="primary-button">
          Click here
        </Button>
      </ThemeProvider>,
    );

    expect(getByTestId('primary-button')).toHaveStyleRule(
      'background-color',
      '#90caf9',
    );
    expect(getByTestId('primary-button')).toHaveStyleRule(
      'border-color',
      'transparent',
    );
  });

  test('should render secondary default button', () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <Button
          buttonType={ButtonAction.SECONDARY}
          size={ButtonSize.DEFAULT}
          onClick={onClick}
          data-testid="secondary-button">
          Click here
        </Button>
      </ThemeProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  test('should render secondary button in disabled mode', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <Button
          buttonType={ButtonAction.SECONDARY}
          size={ButtonSize.DEFAULT}
          onClick={onClick}
          data-testid="secondary-button"
          disabled>
          Click here
        </Button>
      </ThemeProvider>,
    );

    expect(getByTestId('secondary-button')).toHaveStyleRule(
      'background-color',
      '#ffffff',
    );
    expect(getByTestId('secondary-button')).toHaveStyleRule(
      'border-color',
      '#90caf9',
    );
  });
});
