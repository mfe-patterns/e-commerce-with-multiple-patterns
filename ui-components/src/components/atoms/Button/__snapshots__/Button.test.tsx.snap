// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render primary default button 1`] = `
.c0 {
  width: 100%;
  cursor: pointer;
  min-height: 46px;
  border: none;
  border-radius: 23px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: transparent;
  background-color: #2196f3;
  padding: 12px 25.5px;
  color: #ffffff;
}

<div>
  <button
    class="c0"
    data-testid="primary-button"
  >
    Click here
  </button>
</div>
`;

exports[`Button should render secondary default button 1`] = `
.c0 {
  width: 100%;
  cursor: pointer;
  min-height: 46px;
  border: 1px solid;
  border-radius: 23px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-color: #2196f3;
  background-color: #ffffff;
  padding: 12px 25.5px;
  color: #111111;
}

<div>
  <button
    class="c0"
    data-testid="secondary-button"
  >
    Click here
  </button>
</div>
`;
